README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mSigTools

<!-- badges: start -->

[![R-CMD-check](https://github.com/Rozen-Lab/mSigTools/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/Rozen-Lab/mSigTools/actions/workflows/R-CMD-check.yaml)
[![CRAN
status](https://www.r-pkg.org/badges/version/mSigTools)](https://CRAN.R-project.org/package=mSigTools)
[![License: GPL
v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
<!-- badges: end -->

## Purpose

Utility functions for mutational signature analysis.

## Installation

### Get the latest released version

``` r
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}
remotes::install_github(repo = "Rozen-Lab/mSigTools")
```

### Get the development version

Install mSigTools from the master branch on
[GitHub](https://github.com/):

``` r
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}
remotes::install_github(repo = "Rozen-Lab/mSigTools", ref = "master")
```

## Reference manual

<https://github.com/Rozen-Lab/mSigTools/blob/v1.0.8-branch/mSigTools_1.0.8.pdf>


NEWS.md
# mSigTools 1.0.8
* Fixed a bug in function `plot_exposure_internal` when user specifies argument `density`.
* Enabled user to specify argument `border` in exposure plotting functions.

# mSigTools 1.0.7 (submitted to CRAN)
* Renamed exported function `best_reconstruction_QP` to `find_best_reconstruction_QP`.

# mSigTools 1.0.6
* Added function `optimize_exposure_QP`.
* Added `best_reconstruction_QP`.
* Fixed a bug in function `sig_dist_matrix` when in some very rare cases, some
  cosine similarities can be slightly greater than 1.
* Fixed a bug in function `plot_exposure_internal` not to plot sample names when
there are no column names in`exposure`.
* Removed package `PCAWG7` from Suggests field in DESCRIPTION. Added `ICAMS` as suggested package.
* Added new dependency package `sets` as it is required in function `best_reconstruction_QP`.

# mSigTools 1.0.5 (submitted to CRAN)
* Updated test functions in tests/testthat/ folder.

# mSigTools 1.0.4
* Fixed test in test_exposure_related_functions.R.

# mSigTools 1.0.3
* Added return value for exported functions `match_two_sig_sets` and
`write_exposure`.

# mSigTools 1.0.2
* Added reference for "Hungarian algorithm" in package DESCRIPTION.

# mSigTools 1.0.1
* Updated documentation for function `TP_FP_FN_avg_sim`.

# mSigTools 1.0.0
* First stable release.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(mSigTools)

test_check("mSigTools")


tests/testthat/test_sig_dist_matrix.R
test_that("test sig_dist_matrix", {
  ex.sigs <- matrix(c(0.2, 0.8, 0.3, 0.7), nrow = 2)
  colnames(ex.sigs) <- c("ex1", "ex2")
  ref.sigs <- matrix(c(0.21, 0.79, 0.19, 0.81), nrow = 2)
  colnames(ref.sigs) <- c("ref1", "ref2")
  rr <- sig_dist_matrix(ex.sigs, ref.sigs)
  expect_equal(
    rr,
    structure(c(
      0.999889952921527,
      0.989494716817794,
      0.999893768683133,
      0.98481531879193
    ),
    .Dim = c(2L, 2L),
    .Dimnames = list(c("ex1", "ex2"), c("ref1", "ref2"))
    )
  )
  ex.sigs2 <- cbind(ex.sigs, ex3 = c(0.18, 0.82))
  rr <- sig_dist_matrix(ex.sigs2, ref.sigs)
  testthat::expect_equal(
    rr,
    structure(c(
      0.999889952921527,
      0.989494716817794,
      0.999044022249416,
      0.999893768683133,
      0.98481531879193,
      0.999897506866566
    ),
    .Dim = 3:2,
    .Dimnames = list(c("ex1", "ex2", "ex3"), c("ref1", "ref2"))
    )
  )
})


tests/testthat/test_TP_FP_FN_avg_sim.R
test_that("test TP_FP_FN_avg_sim", {
  ex.sigs <- matrix(c(0.2, 0.8, 0.3, 0.7), nrow = 2)
  colnames(ex.sigs) <- c("ex1", "ex2")
  ref.sigs <- matrix(c(0.21, 0.79, 0.19, 0.81), nrow = 2)
  colnames(ref.sigs) <- c("ref1", "ref2")
  rr <- TP_FP_FN_avg_sim(
    extracted.sigs    = ex.sigs,
    reference.sigs    = ref.sigs,
    similarity.cutoff = .985
  )
  expect_equal(
    rr,
    list(TP = 2L,
         FP = 0L,
         FN = 0L,
         avg.cos.sim = 0.994694242750464,
         table = structure(
           list(ex.sig = c("ex1", "ex2"),
                ref.sig = c("ref2",      "ref1"),
                sim = c(0.999893768683133, 0.989494716817794)),
           row.names = 1:2, class = "data.frame"),
         sim.matrix = structure(c(0.999889952921527, 0.989494716817794,
                                  0.999893768683133, 0.98481531879193),
                                .Dim = c(2L, 2L),
                                .Dimnames = list(
                                  c("ex1", "ex2"), c("ref1", "ref2"))),
         unmatched.ex.sigs = character(0),
         unmatched.ref.sigs = character(0))
  )

  ex.sigs2 <- cbind(ex.sigs, ex3 = c(0.18, 0.82))
  rr <- TP_FP_FN_avg_sim(
    extracted.sigs = ex.sigs2,
    reference.sigs = ref.sigs,
    similarity.cutoff = .985
  )
  expect_equal(
    rr,
    list(TP = 2L, FP = 1L, FN = 0L,
         avg.cos.sim = 0.999893729894046,
         table = structure(
           list(ex.sig = c("ex1", "ex3"), ref.sig = c("ref1",      "ref2"),
                sim = c(0.999889952921527, 0.999897506866566)),
           row.names = 1:2, class = "data.frame"),
         sim.matrix = structure(
           c(0.999889952921527, 0.989494716817794,
             0.999044022249416, 0.999893768683133,
             0.98481531879193, 0.999897506866566     ),
           .Dim = 3:2,
           .Dimnames = list(c("ex1", "ex2", "ex3"), c("ref1",      "ref2"))),
         unmatched.ex.sigs = "ex2",
         unmatched.ref.sigs = character(0))
  )
})


R/sig_dist_matrix.R
#' Compute a matrix of distances / similarities between two sets of signatures.
#'
#' @param x1 The first set of signatures (a numerical matrix-like object
#'     in which each column is a signature).
#'
#' @param x2 The second set of signatures, similar data type to \code{x1},
#'     and must have the same number of rows as \code{x1}.
#'
#' @param method As for the \code{\link[philentropy]{distance}} function in package
#'   \code{philenropy}.
#'
#' @return A numeric matrix with dimensions
#'   \code{ncol(x1)} X \code{ncol(x2)}.
#'   Each element represents the distance or
#'   similarity (depending on \code{method})
#'   between a column in \code{x1} and a column in \code{x2}.
#'
#' @export
#'
#' @examples
#' ex.sigs <- matrix(c(0.2, 0.8, 0.3, 0.7, 0.4, 0.6), nrow = 2)
#' colnames(ex.sigs) <- c("ex1", "ex2", "ex3")
#' ref.sigs <- matrix(c(0.21, 0.79, 0.19, 0.81), nrow = 2)
#' colnames(ref.sigs) <- c("ref1", "ref2")
#' sig_dist_matrix(ex.sigs, ref.sigs)
#'
sig_dist_matrix <- function(x1, x2, method = "cosine") {
  mm <- cbind(x1, x2)
  dd <- suppressMessages(
    philentropy::distance(t(mm), method = method, use.row.names = TRUE)
  )

  # In some very rare cases, some cosine similarities in dd may be slightly
  # greater than 1 e.g. 1.00000000000000022204460492503130808472633361816406250
  # To avoid error later, we change those values to 1
  if (method == "cosine") {
    dd[dd > 1] <- 1
  }

  dd2 <- dd[1:ncol(x1), , drop = FALSE] # Use the rows that represent the elements of x1
  dd3 <- dd2[, -(1:ncol(x1)), drop = FALSE] # Use the columns that represent elements of x2
  return(dd3)
}


R/internal_matches.R
# We break this out as a separate internal function to simplify testing.
internal_matches <- function(original.dd, cutoff) {

  # Cannot use Inf in the foreign function call clue::solve_LSAP(dd) (below)
  my.inf <- 9e99

  # browser()

  dd <- original.dd
  was.transformed <- FALSE
  if (nrow(dd) > ncol(dd)) {
    # Add more columns
    dd <- t(dd)
    was.transformed <- TRUE
  }

  dd[dd > cutoff] <- my.inf

  solution <- clue::solve_LSAP(dd)

  cost.one.pair <- function(rowi) {
    # browser()
    colj <- solution[rowi]
    dist <- dd[rowi, colj]
    name1 <- rownames(dd)[rowi]
    name2 <- colnames(dd)[colj]
    return(list(name1 = name1, name2 = name2, dist = dist))
  }

  proto.table <- lapply(1:length(solution), FUN = cost.one.pair)

  table1 <- matrix(unlist(proto.table), ncol = 3, byrow = TRUE)
  if (was.transformed) {
    table1 <- table1[, c(2, 1, 3)]
    dd <- t(dd)
  }

  colnames(table1) <- c("x1", "x2", "dist")
  # browser()
  table1 <- data.frame(table1)
  table1$dist <- as.numeric(table1$dist)

  table <- table1[table1$dist < my.inf, , drop = FALSE]

  return(list(table = table, orig.matrix = original.dd, modified.matrix = dd))
}


R/match_two_sig_sets.R
#' Find an optimal matching between two sets of signatures subject to a maximum distance.
#'
#' @param x1 A numerical-matrix-like object with columns as signatures.
#'
#' @param x2 A numerical-matrix-like object with columns as signatures.
#'   Needs to have the same number of rows as \code{x1}.
#'
#' @param method As for the \code{\link[philentropy]{distance}} function in package
#'   \code{philenropy}.
#'
#' @param convert.sim.to.dist If \code{method} specifies a similarity
#'   rather than a distance, use this function to convert the
#'   similarity to a distance.
#'
#' @param cutoff A maximum distance or minimum similarity over which to
#'   pair signatures between \code{x1} and \code{x2}.
#'
#' @details Match signatures between \code{x1} and \code{x2}
#'     using the function
#'    \code{\link[clue]{solve_LSAP}}, which uses the
#'    "Hungarian" (a.k.a "Kuhn–Munkres") algorithm
#'    \url{https://en.wikipedia.org/wiki/Hungarian_algorithm},
#'    which optimizes the total cost associated with the links
#'    between nodes.
#'    This function generates a distance matrix between the two
#'    sets of signatures using \code{method} and, if necessary,
#'    \code{convert.sim.to.dist}.
#'    It then sets distances > \code{cutoff} to very large values and
#'    then applies \code{\link[clue]{solve_LSAP}} to the resulting
#'    matrix to compute a matching between
#'    \code{x1} and \code{x2} that minimizes the sum of the
#'    distances.
#'
#' @return A list with the elements
#'
#' * \code{table} Table of extracted signatures that matched a reference
#'    signature. Each row contains the extracted signature name,
#'    the reference
#'    signature name, and the distance of the match.
#'
#' * \code{orig.matrix} The matrix of numeric distances between
#'     \code{x1} and \code{x2}.
#'
#' * \code{modified.matrix} The argument \code{orig.matrix}
#'    with distances >
#'    \code{cutoff} changed to very large values.
#'
#' @export
#'
#' @md
#'
#' @examples
#' ex.sigs <- matrix(c(0.2, 0.8, 0.3, 0.7, 0.6, 0.4), nrow = 2)
#' colnames(ex.sigs) <- c("ex1", "ex2", "ex3")
#' ref.sigs <- matrix(c(0.21, 0.79, 0.19, 0.81), nrow = 2)
#' colnames(ref.sigs) <- c("ref1", "ref2")
#' match_two_sig_sets(ex.sigs, ref.sigs, cutoff = .9)
#'
match_two_sig_sets <-
  function(x1,
           x2,
           method = "cosine",
           convert.sim.to.dist = function(x) {
             return(1 - x)
           },
           cutoff = 0.9) {
    # browser()
    dd <- sig_dist_matrix(x1, x2, method = method)
    if (!is.null(convert.sim.to.dist)) {
      dd <- convert.sim.to.dist(dd)
      cutoff <- convert.sim.to.dist(cutoff)
    }

    return(internal_matches(dd, cutoff))
  }


R/optimize_exposure_QP.R
#' Quadratic programming optimization of signature activities
#'
#' @param spectrum Mutational signature or mutational spectrum
#'      as a numeric vector or single column data frame or matrix.
#'
#' @param signatures Matrix or data frame of signatures from which to
#'   reconstruct \code{spectrum}. Rows are mutation types and columns are
#'   signatures. Should have column names for interpretable results. Cannot be a
#'   vector because the column names are needed.
#'
#' @details
#' Code adapted from \code{SignatureEstimation::decomposeQP} and
#' uses \code{\link[quadprog]{solve.QP}} in package \code{quadprog}.
#'
#' @export
#'
#' @return A vector of exposures with names being the \code{colnames} from
#'   \code{signatures}.
#'
#' @examples
#' usigs <- matrix(c(0.2, 0.7, 0.1,
#'                   0.3, 0.6, 0.1,
#'                   0.1, 0.1, 0.8), nrow = 3)
#' colnames(usigs) <- c("u1", "u2", "u3")
#' tsig <- matrix(c(0.25, 0.65, 0.1), nrow = 3)
#' optimize_exposure_QP(tsig, usigs)

optimize_exposure_QP <- function(spectrum, signatures) {
  if (is.null(spectrum)) {
    traceback()
    stop("Got spectrum = NULL")
  }

  if (is.data.frame(signatures)) {
    signatures <- as.matrix(signatures)
  }
  stopifnot(is.matrix(signatures))

  if (sum(spectrum) == 0 ) {
    rr <- rep(0, ncol(signatures))
    names(rr) <- colnames(signatures)
    return(rr)
  }

  M <- spectrum / sum(spectrum)

  P <- signatures

  # N: how many signatures should be considered
  N = ncol(P)
  if (ncol(P) == 1) {
    rr <- sum(spectrum)
    names(rr) <- colnames(P)
    stopifnot(!is.null(names(rr)))
    return(rr)
  }

  # Matrix appearing in the quadratic programming objective function
  G = t(P) %*% P

  # Constraints under which to minimize the objective function
  C <- cbind(rep(1,N), diag(N))

  # b: vector containing the values of b_0.
  b <- c(1,rep(0,N))

  # d: vector appearing in the quatric programming objective function
  d <- t(M) %*% P

  out <- quadprog::solve.QP(Dmat = G, dvec = d, Amat = C, bvec = b, meq = 1)

  # Some exposure values may be < but very close to 0;
  # Change these to 0 and renormalize

  exposures <- out$solution
  names(exposures) <- colnames(signatures)
  exposures[exposures < 0] <- 0

  rr <- sum(spectrum) * exposures/sum(exposures)
  stopifnot(!is.null(names(rr)))
  return(rr)
}


R/best_reconstruction_QP.R
#' Find "best" reconstruction of a target signature or spectrum from a set of signatures.
#'
#' @param target.sig The signature or spectrum to reconstruct;
#'   a non-negative numeric vector or 1-column matrix-like object.
#'
#' @param sig.universe The universe of signatures from which to reconstruct
#'   \code{target.sig}; a non-negative numeric matrix-like object with
#'   \code{nrow(sig.universe) == length(target.sig)}. The sums of each column
#'   must be 1. Must not contain duplicate columns or have other
#'   non-unique quadratic programming solutions (not checked, but will generate
#'   an error from \code{\link[quadprog]{solve.QP}} in package `quadprog`).
#'
#' @param max.subset.size Maximum number of signatures to use to
#'   reconstruct \code{target.sig}.
#'
#' @param method As in \code{\link[philentropy]{dist_one_one}} in
#'   package \code{philentropy}, and used only
#'   to find the final "best" reconstruction. The optimized exposures from which
#'   to selected the "best" reconstruction are calculated using
#'   \code{\link{optimize_exposure_QP}}, which uses \code{\link[quadprog]{solve.QP}}
#'   in package \code{quadprog}.
#'
#' @param trim.less.than After optimizing exposures with
#'   \code{\link{optimize_exposure_QP}}, remove exposures less than
#'   \code{trim.less.than} and then re-optimize.
#'
#' @md
#'
#' @return A list with elements:
#'
#' * \code{optimized.exposure} A numerical vector of the exposures that
#'   give the "best" reconstruction. This vector is empty if there is
#'   an error.
#'
#' * \code{similarity} The similarity between the \code{reconstruction}
#'   (see below) and \code{target.sig} according to the distance
#'   or similarity provided by the \code{method} argument.
#'
#' * \code{method} The value specified for the `method` argument,
#'   or an error message if `optimize.exposure` is empty.
#'
#' * \code{reconstruction} The reconstruction of `target.sig` according to
#'   `optimized.exposure`.
#'
#' @export
#'
#' @details
#'   This function should be fast if you do not specify \code{max.subset.size},
#'   but it will be combinatorially slow if \code{max.subset.size} is large
#'   and \code{trim.less.than} is small or negative. So do not do that.
#'   If `max.subset.size` is `NULL`, then the function just uses \code{\link{optimize_exposure_QP}}.
#'   and then excludes exposures < `trim.less.than`, and then re-runs
#'   \code{\link{optimize_exposure_QP}}. Otherwise, after excluding
#'   exposures < `trim.less.than`, then the function runs \code{\link{optimize_exposure_QP}} on
#'   subsets of signatures of size <= `max.subset.size`, removes exposures < `trim.less.than`,
#'   reruns \code{\link{optimize_exposure_QP}}, calculates the reconstruction and
#'   similarity between the reconstruction and the `target.sig` and returns the information for
#'   the exposures that have the greatest similarity.
#'
#' @examples
#' set.seed(888)
#'   sig.u <-
#' do.call(cbind,
#'        lapply(1:6, function(x) { col <- runif(n = 96); col/sum(col) }))
#' rr <- best_reconstruction_QP(target.sig = sig.u[ , 1, drop = FALSE],
#'                             sig.universe = sig.u[ , 2:6])
#' names(rr)
#' rr$optimized.exposure
#' rr$similarity
#' rr <- best_reconstruction_QP(target.sig = sig.u[ , 1, drop = FALSE],
#'                             sig.universe = sig.u[ , 2:6],
#'                             max.subset.size = 3)
#' rr$optimized.exposure
#' rr$similarity

best_reconstruction_QP <- function(target.sig,
                                   sig.universe,
                                   max.subset.size = NULL,
                                   method          = "cosine",
                                   trim.less.than  = 1e-10) {

  if (!isTRUE(
    all.equal.numeric(colSums(sig.universe),
                      rep(1, ncol(sig.universe)),
                      check.attributes = FALSE)
    )) {
    return(list(optimized.exposure = numeric(),
                method = "Error: all signature.universe columns must sum to 1"))
  }

  if (is.null(colnames(sig.universe))) {
    colnames(sig.universe) <- paste("v", 1:ncol(sig.universe), sep = "")
  }

  optimize.etc <- function(target.sig, sig.u) {
    optimized.exposure <- optimize_exposure_QP(target.sig, sig.u)
    to.remove <- which(optimized.exposure < trim.less.than)
    if (length(to.remove) > 1) {
      optimized.exposure <- optimized.exposure[-to.remove]
    }
    used.sigs <- sig.u[ , names(optimized.exposure), drop = FALSE]
    if (length(to.remove) > 1) {
      optimized.exposure <- optimize_exposure_QP(target.sig, used.sigs)
    }

    recon <- used.sigs %*% optimized.exposure

    dist <- philentropy::dist_one_one(target.sig, recon, method = method)

    return(list(optimized.exposure = optimized.exposure,
                similarity         = dist,
                method             = method,
                reconstruction     = recon))
  }

  result1 <- optimize.etc(target.sig, sig.universe)

  if (is.null(max.subset.size)) {
    return(result1)
  }

  cnames <- names(result1[["optimized.exposure"]])

  if (length(cnames) < max.subset.size) {
    max.subset.size <-length(cnames)
  }
  xx <- lapply(1:max.subset.size,
               function(m) {
                 xxx <- sets::gset_combn(cnames,  m = m)
                 return(xxx)
               })
  xx <- do.call(sets::gset_union, xx)

  one.sig.set <- function(sig.name.set) {
    # browser()
    rrr <- optimize.etc(
      target.sig,
      sig.universe[ , as.character(sig.name.set), drop = FALSE])
    return(rrr)
  }

  rr <- lapply(X = xx, FUN = one.sig.set)
  similarities <- unlist(lapply(rr, `[[`, "similarity"))
  best <- which(similarities == max(similarities))
  return(rr[[best]])
}


R/find_best_reconstruction_QP.R
#' Find "best" reconstruction of a target signature or spectrum from a set of signatures.
#'
#' @param target.sig The signature or spectrum to reconstruct;
#'   a non-negative numeric vector or 1-column matrix-like object.
#'
#' @param sig.universe The universe of signatures from which to reconstruct
#'   \code{target.sig}; a non-negative numeric matrix-like object with
#'   \code{nrow(sig.universe) == length(target.sig)}. The sums of each column
#'   must be 1. Must not contain duplicate columns or have other
#'   non-unique quadratic programming solutions (not checked, but will generate
#'   an error from \code{\link[quadprog]{solve.QP}} in package `quadprog`).
#'
#' @param max.subset.size Maximum number of signatures to use to
#'   reconstruct \code{target.sig}.
#'
#' @param method As in \code{\link[philentropy]{dist_one_one}} in
#'   package \code{philentropy}, and used only
#'   to find the final "best" reconstruction. The optimized exposures from which
#'   to selected the "best" reconstruction are calculated using
#'   \code{\link{optimize_exposure_QP}}, which uses \code{\link[quadprog]{solve.QP}}
#'   in package \code{quadprog}.
#'
#' @param trim.less.than After optimizing exposures with
#'   \code{\link{optimize_exposure_QP}}, remove exposures less than
#'   \code{trim.less.than} and then re-optimize.
#'
#' @md
#'
#' @return A list with elements:
#'
#' * \code{optimized.exposure} A numerical vector of the exposures that
#'   give the "best" reconstruction. This vector is empty if there is
#'   an error.
#'
#' * \code{similarity} The similarity between the \code{reconstruction}
#'   (see below) and \code{target.sig} according to the distance
#'   or similarity provided by the \code{method} argument.
#'
#' * \code{method} The value specified for the `method` argument,
#'   or an error message if `optimize.exposure` is empty.
#'
#' * \code{reconstruction} The reconstruction of `target.sig` according to
#'   `optimized.exposure`.
#'
#' @export
#'
#' @details
#'   This function should be fast if you do not specify \code{max.subset.size},
#'   but it will be combinatorially slow if \code{max.subset.size} is large
#'   and \code{trim.less.than} is small or negative. So do not do that.
#'   If `max.subset.size` is `NULL`, then the function just uses \code{\link{optimize_exposure_QP}}.
#'   and then excludes exposures < `trim.less.than`, and then re-runs
#'   \code{\link{optimize_exposure_QP}}. Otherwise, after excluding
#'   exposures < `trim.less.than`, then the function runs \code{\link{optimize_exposure_QP}} on
#'   subsets of signatures of size <= `max.subset.size`, removes exposures < `trim.less.than`,
#'   reruns \code{\link{optimize_exposure_QP}}, calculates the reconstruction and
#'   similarity between the reconstruction and the `target.sig` and returns the information for
#'   the exposures that have the greatest similarity.
#'
#' @examples
#' set.seed(888)
#' sig.u <-
#'   do.call(
#'     cbind,
#'     lapply(1:6, function(x) {
#'       col <- runif(n = 96)
#'       col / sum(col)
#'     })
#'   )
#' rr <- find_best_reconstruction_QP(
#'   target.sig = sig.u[, 1, drop = FALSE],
#'   sig.universe = sig.u[, 2:6]
#' )
#' names(rr)
#' rr$optimized.exposure
#' rr$similarity
#' rr <- find_best_reconstruction_QP(
#'   target.sig = sig.u[, 1, drop = FALSE],
#'   sig.universe = sig.u[, 2:6],
#'   max.subset.size = 3
#' )
#' rr$optimized.exposure
#' rr$similarity
#'
find_best_reconstruction_QP <- function(target.sig,
                                        sig.universe,
                                        max.subset.size = NULL,
                                        method = "cosine",
                                        trim.less.than = 1e-10) {
  if (!isTRUE(
    all.equal.numeric(colSums(sig.universe),
      rep(1, ncol(sig.universe)),
      check.attributes = FALSE
    )
  )) {
    return(list(
      optimized.exposure = numeric(),
      method = "Error: all signature.universe columns must sum to 1"
    ))
  }

  if (is.null(colnames(sig.universe))) {
    colnames(sig.universe) <- paste("v", 1:ncol(sig.universe), sep = "")
  }

  optimize.etc <- function(target.sig, sig.u) {
    optimized.exposure <- optimize_exposure_QP(target.sig, sig.u)
    to.remove <- which(optimized.exposure < trim.less.than)
    if (length(to.remove) > 1) {
      optimized.exposure <- optimized.exposure[-to.remove]
    }
    used.sigs <- sig.u[, names(optimized.exposure), drop = FALSE]
    if (length(to.remove) > 1) {
      optimized.exposure <- optimize_exposure_QP(target.sig, used.sigs)
    }

    recon <- used.sigs %*% optimized.exposure

    dist <- philentropy::dist_one_one(target.sig, recon, method = method)

    return(list(
      optimized.exposure = optimized.exposure,
      similarity = dist,
      method = method,
      reconstruction = recon
    ))
  }

  result1 <- optimize.etc(target.sig, sig.universe)

  if (is.null(max.subset.size)) {
    return(result1)
  }

  cnames <- names(result1[["optimized.exposure"]])

  if (length(cnames) < max.subset.size) {
    max.subset.size <- length(cnames)
  }
  xx <- lapply(
    1:max.subset.size,
    function(m) {
      xxx <- sets::gset_combn(cnames, m = m)
      return(xxx)
    }
  )
  xx <- do.call(sets::gset_union, xx)

  one.sig.set <- function(sig.name.set) {
    # browser()
    rrr <- optimize.etc(
      target.sig,
      sig.universe[, as.character(sig.name.set), drop = FALSE]
    )
    return(rrr)
  }

  rr <- lapply(X = xx, FUN = one.sig.set)
  similarities <- unlist(lapply(rr, `[[`, "similarity"))
  best <- which(similarities == max(similarities))
  return(rr[[best]])
}


R/TP_FP_FN_avg_sim.R
#' Find best matches (by cosine similarity) of a set of mutational signatures to
#' a set of reference mutational signatures.
#'
#' @details Match signatures in \code{extracted.sigs} to
#'    signatures in \code{reference.sigs} using \code{\link{match_two_sig_sets}}
#'    based on cosine similarity.
#'
#' @param extracted.sigs Mutational signatures discovered by some analysis.
#'    A numerical-matrix-like object with columns as signatures.
#'
#' @param reference.sigs A numerical-matrix-like object with columns
#'    as signatures. This matrix should contain the
#'    reference mutational signatures. For example,
#'    these might be from a synthetic data set or they
#'    could be from reference set of signatures,
#'    such as the signatures at the COSMIC mutational
#'    signatures web site. See CRAN package cosmicsig.
#'
#' @param similarity.cutoff A signature in \code{reference.sigs}
#'    must be matched
#'    by \code{>= similarity.cutoff} by a signature in \code{extracted.sigs}
#'    to be considered detected.
#'
#' @export
#'
#' @return A list with the elements
#'
#' * \code{TP} The number of true positive extracted signatures.
#'
#' * \code{FP} The number of false positive extracted signatures.
#'
#' * \code{FN} The number of false negative reference signatures.
#'
#' * \code{avg.cos.sim} The average cosine similarity of
#'     true positives to their matching reference signatures.
#'
#' * \code{table} A data.frame of extracted signatures
#'     that matched a reference signature.
#'     Each row contains the extracted signature name,
#'     the reference signature name, and the
#'     cosine similarity of the match.
#'
#' * \code{sim.matrix} The numeric distance or similarity
#'     matrix between \code{extracted.sigs} and
#'     \code{reference.sigs} as returned from
#'     \code{\link{sig_dist_matrix}}.
#'
#' * \code{unmatched.ex.sigs} The identifiers of
#'     the extracted signatures that did not match a
#'     reference signature.
#'
#' * \code{unmatched.ref.sigs} The identifiers of
#'     the reference signatures that did not match an
#'     extracted signature.
#'
#' @examples
#' ex.sigs <- matrix(c(0.2, 0.8, 0.3, 0.7, 0.6, 0.4), nrow = 2)
#' colnames(ex.sigs) <- c("ex1", "ex2", "ex3")
#' ref.sigs <- matrix(c(0.21, 0.79, 0.19, 0.81), nrow = 2)
#' colnames(ref.sigs) <- c("ref1", "ref2")
#' TP_FP_FN_avg_sim(
#'   extracted.sigs = ex.sigs,
#'   reference.sigs = ref.sigs,
#'   similarity.cutoff = .9
#' )
#'
#' @md
#'
TP_FP_FN_avg_sim <-
  function(extracted.sigs, reference.sigs, similarity.cutoff = 0.9) {
    tt.and.matrix <-
      match_two_sig_sets(
        extracted.sigs,
        reference.sigs,
        method = "cosine",
        cutoff = similarity.cutoff
      )
    # browser()
    tt <- tt.and.matrix$table
    TP.sigs <- tt[, 1]
    FP.sigs <- setdiff(colnames(extracted.sigs), TP.sigs)
    FN.sigs <- setdiff(colnames(reference.sigs), tt[, 2])
    tt[, 3] <- 1 - tt[, 3]
    colnames(tt) <- c("ex.sig", "ref.sig", "sim")
    return(list(
      TP = length(TP.sigs),
      FP = length(FP.sigs),
      FN = length(FN.sigs),
      avg.cos.sim = mean(tt[, 3]),
      table = tt,
      sim.matrix = 1 - tt.and.matrix$orig.matrix,
      unmatched.ex.sigs = FP.sigs,
      unmatched.ref.sigs = FN.sigs
    ))
  }
